crates/scroll/evm/src/config.rs

```diff
@@ -0,0 +1,326 @@
+    chain_spec: Arc<ScrollChainSpec>,
+    /// Additional Scroll configuration.
+    scroll_config: ScrollChainConfig,
+}
+
+impl ScrollEvmConfig {
+    /// Returns a new instance of [`ScrollEvmConfig`].
+    pub const fn new(chain_spec: Arc<ScrollChainSpec>, scroll_config: ScrollChainConfig) -> Self {
+        Self { chain_spec, scroll_config }
+    }
+
+    /// Returns the spec id at the given head.
+    pub fn spec_id_at_head(&self, head: &Head) -> SpecId {
+        let chain_spec = &self.chain_spec;
+        if chain_spec.fork(ScrollHardfork::Curie).active_at_head(head) {
+            SpecId::CURIE
+        } else if chain_spec.fork(ScrollHardfork::Bernoulli).active_at_head(head) {
+            SpecId::BERNOULLI
+        } else {
+            SpecId::PRE_BERNOULLI
+        }
+    }
+}
+
+impl ConfigureEvm for ScrollEvmConfig {
+    type DefaultExternalContext<'a> = ();
+
+    fn evm<DB: Database>(&self, db: DB) -> Evm<'_, Self::DefaultExternalContext<'_>, DB> {
+        EvmBuilder::default().with_db(db).scroll().build()
+    }
+
+    fn evm_with_inspector<DB, I>(&self, db: DB, inspector: I) -> Evm<'_, I, DB>
+    where
+        DB: Database,
+        I: GetInspector<DB>,
+    {
+        EvmBuilder::default()
+            .with_db(db)
+            .with_external_context(inspector)
+            .scroll()
+            .append_handler_register(inspector_handle_register)
+            .build()
+    }
+
+    fn default_external_context<'a>(&self) -> Self::DefaultExternalContext<'a> {}
+}
+
+impl ConfigureEvmEnv for ScrollEvmConfig {
+    type Header = alloy_consensus::Header;
+    type Error = Infallible;
+
+    fn fill_tx_env(&self, tx_env: &mut TxEnv, transaction: &TransactionSigned, sender: Address) {
+        transaction.fill_tx_env(tx_env, sender);
+    }
+
+    fn fill_tx_env_system_contract_call(
+        &self,
+        _env: &mut Env,
+        _caller: Address,
+        _contract: Address,
+        _data: Bytes,
+    ) {
+        /* noop */
+    }
+
+    fn fill_cfg_env(
+        &self,
+        cfg_env: &mut CfgEnvWithHandlerCfg,
+        header: &Self::Header,
+        total_difficulty: U256,
+    ) {
+        let spec_id = self.spec_id_at_head(&Head {
+            number: header.number,
+            timestamp: header.timestamp,
+            difficulty: header.difficulty,
+            total_difficulty,
+            ..Default::default()
+        });
+
+        cfg_env.handler_cfg.spec_id = spec_id;
+        cfg_env.handler_cfg.is_scroll = true;
+
+        cfg_env.chain_id = self.chain_spec.chain().id();
+        cfg_env.perf_analyse_created_bytecodes = AnalysisKind::Analyse;
+    }
+
+    fn fill_block_env(&self, block_env: &mut BlockEnv, header: &Self::Header, after_merge: bool) {
+        block_env.number = U256::from(header.number);
+        block_env.coinbase = header.beneficiary;
+        if let Some(vault_address) = self.scroll_config.fee_vault_address {
+            block_env.coinbase = vault_address;
+        }
```

Comment:
Should we set the `block_env.coinbase` using:

```rust
        if let Some(vault_address) = self.scroll_config.fee_vault_address {
            block_env.coinbase = vault_address;
        } else {
            block_env.coinbase = header.beneficiary;
        }
```

This should prevent the additional assignment on line 108.